backends/bmv2/common/control.h

```diff
@@ -39,18 +48,777 @@ class ControlConverter : public Inspector {
+            if (implementationType->name == actionSelectorName) {
+                BUG_CHECK(arguments->size() == 3, "%1%: expected 3 arguments", arguments);
+                isSimpleTable = false;
+                auto selector = new Util::JsonObject();
+                table->emplace("type", "indirect_ws");
+                action_profile->emplace("selector", selector);
+                add_size(1);
+                auto hash = arguments->at(0)->expression;
+                auto ei = P4::EnumInstance::resolve(hash, ctxt->typeMap);
+                if (ei == nullptr) {
+                    ::error(ErrorType::ERR_EXPECTED,
+                            "%1%: hash must be a constant on this target", hash);
+                } else {
+                    cstring algo = ei->name;
+                    selector->emplace("algo", algo);
+                }
+                auto input = mkArrayField(selector, "input");
+                for (auto ke : key->keyElements) {
+                    auto mt = ctxt->refMap->getDeclaration(ke->matchType->path, true)
+                            ->to<IR::Declaration_ID>();
+                    BUG_CHECK(mt != nullptr, "%1%: could not find declaration", ke->matchType);
+                    if (mt->name.name != BMV2::MatchImplementation::selectorMatchTypeName)
+                        continue;
+
+                    auto expr = ke->expression;
+                    auto jk = ctxt->conv->convert(expr);
+                    input->append(jk);
+                }
+            } else if (implementationType->name ==
+                    Standard::ActionProfileTraits<arch>::typeName()) {
+                isSimpleTable = false;
+                table->emplace("type", "indirect");
+                add_size(0);
+            } else {
+                ::error(ErrorType::ERR_UNEXPECTED, "%1%: expected value for property", propv);
+            }
+        } else if (propv->expression->is<IR::PathExpression>()) {
+            auto pathe = propv->expression->to<IR::PathExpression>();
+            auto decl = ctxt->refMap->getDeclaration(pathe->path, true);
+            if (!decl->is<IR::Declaration_Instance>()) {
+                ::error(ErrorType::ERR_EXPECTED, "%1%: expected a reference to an instance", pathe);
+                return false;
+            }
+            apname = decl->controlPlaneName();
+            auto dcltype = ctxt->typeMap->getType(pathe, true);
+            if (!dcltype->is<IR::Type_Extern>()) {
+                ::error(ErrorType::ERR_UNEXPECTED, "%1%: unexpected type for implementation", dcltype);
+                return false;
+            }
+            auto type_extern_name = dcltype->to<IR::Type_Extern>()->name;
+            auto actionProfileName = Standard::ActionProfileTraits<arch>::propertyName();
+            auto actionSelectorName = Standard::ActionSelectorTraits<arch>::propertyName();
+            if (type_extern_name == actionProfileName) {
```

Comment:
Used to be `implementationType->name == BMV2::TableImplementation::actionProfileName`